File: bwl.Imaging/Matrix/FloatMatrix/GrayFloatMatrix.vb
Public Class GrayFloatMatrix
    Inherits CommonFloatMatrix

    Sub New(width As Integer, height As Integer)
        MyBase.New(1, width, height)
    End Sub

    Sub New(matrix() As Double, width As Integer, height As Integer)
        MyBase.New({matrix}, width, height)
    End Sub

    Sub New(matrix() As Integer, width As Integer, height As Integer)
        MyBase.New({matrix}, width, height)
    End Sub

    Sub New(matrix() As Double, width As Integer, height As Integer, multiplier As Double)
        MyBase.New({matrix}, width, height, multiplier)
    End Sub

    Sub New(matrix() As Integer, width As Integer, height As Integer, multiplier As Double)
        MyBase.New({matrix}, width, height, multiplier)
    End Sub

    Public Property GrayPixel(x As Integer, y As Integer) As Double
        Get
            Return _matrices(0)(x + y * Width)
        End Get
        Set(value As Double)
            _matrices(0)(x + y * Width) = value
        End Set
    End Property

    Public ReadOnly Property Gray As Double()
        Get
            Return _matrices(0)
        End Get
    End Property

    Public Shadows Function Clone() As GrayFloatMatrix
        Return New GrayFloatMatrix(CloneMatrix(Gray), Width, Height)
    End Function

    Public Function ToRGBFloatMatrix() As RGBFloatMatrix
        Return New RGBFloatMatrix(Gray, Gray, Gray, Width, Height)
    End Function

    Public Overloads Function ResizeTwo() As GrayFloatMatrix
        Dim resized = MyBase.ResizeTwo()
        Return New GrayFloatMatrix(resized.Matrix(0), resized.Width, resized.Height)
    End Function

    Public Overloads Function ResizeHalf() As GrayFloatMatrix
        Dim resized = MyBase.ResizeHalf()
        Return New GrayFloatMatrix(resized.Matrix(0), resized.Width, resized.Height)
    End Function
End Class

File: Bwl.Imaging.Skia/Matrix/FloatMatrix/GrayFloatMatrix.cs
namespace Bwl.Imaging.Skia;

public class GrayFloatMatrix : CommonFloatMatrix
{

    public GrayFloatMatrix(int width, int height) : base(1, width, height)
    {
    }

    public GrayFloatMatrix(double[] matrix, int width, int height) : base(new[] { matrix }, width, height)
    {
    }

    public GrayFloatMatrix(int[] matrix, int width, int height) : base(new[] { matrix }, width, height)
    {
    }

    public GrayFloatMatrix(double[] matrix, int width, int height, double multiplier) : base(new[] { matrix }, width, height, multiplier)
    {
    }

    public GrayFloatMatrix(int[] matrix, int width, int height, double multiplier) : base(new[] { matrix }, width, height, multiplier)
    {
    }

    public double GetGrayPixel(int x, int y)
    {
        return _matrices[0][x + y * Width];
    }
    public void SetGrayPixel(int x, int y, double value)
    {
        _matrices[0][x + y * Width] = value;
    }

    public double[] Gray
    {
        get
        {
            return _matrices[0];
        }
    }

    public new GrayFloatMatrix Clone()
    {
        return new GrayFloatMatrix(CloneMatrix(Gray), Width, Height);
    }

    public RGBFloatMatrix ToRGBFloatMatrix()
    {
        return new RGBFloatMatrix(Gray, Gray, Gray, Width, Height);
    }

    public new GrayFloatMatrix ResizeTwo()
    {
        var resized = base.ResizeTwo();
        return new GrayFloatMatrix(resized.GetMatrix(0), resized.Width, resized.Height);
    }

    public new GrayFloatMatrix ResizeHalf()
    {
        var resized = base.ResizeHalf();
        return new GrayFloatMatrix(resized.GetMatrix(0), resized.Width, resized.Height);
    }
}